hashing/sha256.cpp

```diff
@@ -0,0 +1,227 @@
+
+    // Pre-processing
+    std::string padded_input = input;
+    uint64_t input_size = input.length() * 8;  // Message length in bits
+
+    // Append a single '1' bit
+    padded_input += '\x80';
+
+    // Pad with zeros
+    while ((padded_input.length() * 8) % 512 != 448) padded_input += '\x00';
+
+    // Append length of the original input string as a 64-bit big-endian integer
+    for (uint32_t i = 64; i != 0; i -= 8) {
+        padded_input += static_cast<char>((input_size >> (i - 8)) & 0xFF);
+    }
+
+    // Process message in successive 512-bit (64-byte) chunks
+    for (size_t i = 0; i < padded_input.length(); i += 64) {
+        uint32_t blocks[64];
+
+        // Copy chunk into first 16 words of the message schedule array
+        for (size_t j = 0; j < 16; ++j) {
+            blocks[j] =
+                (static_cast<uint8_t>(padded_input[i + j * 4]) << 24) |
+                (static_cast<uint8_t>(padded_input[i + j * 4 + 1]) << 16) |
+                (static_cast<uint8_t>(padded_input[i + j * 4 + 2]) << 8) |
+                static_cast<uint8_t>(padded_input[i + j * 4 + 3]);
+        }
+
+        for (size_t j = 16; j < 64; ++j) {
+            uint32_t s0 = rightRotate(blocks[j - 15], 7) ^
+                          rightRotate(blocks[j - 15], 18) ^
+                          (blocks[j - 15] >> 3);
+            uint32_t s1 = rightRotate(blocks[j - 2], 17) ^
+                          rightRotate(blocks[j - 2], 19) ^
+                          (blocks[j - 2] >> 10);
+            blocks[j] = blocks[j - 16] + s0 + blocks[j - 7] + s1;
+        }
+
+        // Initialize working variables
+        uint32_t a = hash[0];
+        uint32_t b = hash[1];
+        uint32_t c = hash[2];
+        uint32_t d = hash[3];
+        uint32_t e = hash[4];
+        uint32_t f = hash[5];
+        uint32_t g = hash[6];
+        uint32_t h = hash[7];
+
+        // Compression function main loop
+        for (size_t j = 0; j < 64; ++j) {
+            uint32_t s1 =
+                rightRotate(e, 6) ^ rightRotate(e, 11) ^ rightRotate(e, 25);
+            uint32_t ch = (e & f) ^ (~e & g);
+            uint32_t temp1 = h + s1 + ch + k[j] + blocks[j];
+            uint32_t s0 =
+                rightRotate(a, 2) ^ rightRotate(a, 13) ^ rightRotate(a, 22);
+            uint32_t maj = (a & b) ^ (a & c) ^ (b & c);
+            uint32_t temp2 = s0 + maj;
+
+            h = g;
+            g = f;
+            f = e;
+            e = d + temp1;
+            d = c;
+            c = b;
+            b = a;
+            a = temp1 + temp2;
+        }
+
+        // Update hash values
+        hash[0] += a;
+        hash[1] += b;
+        hash[2] += c;
+        hash[3] += d;
+        hash[4] += e;
+        hash[5] += f;
+        hash[6] += g;
+        hash[7] += h;
+    }
+
+    // Convert the hash to a hexadecimal string
+    std::string result;
+    for (size_t i = 0; i < 8; ++i) {
+        for (int j = 0; j < 4; ++j) {
```

Comment:
Is there a reason that `j` is an `int` and not `size_t`?

Reply:
No, that's a mistake. I'll change it.